File: KadirliApp.Domain/Enums/OperatingDays.cs
using System.Globalization;

namespace KadirliApp.Domain.Enums;

/// <summary>
/// Faz 12.5 — bir seferin <b>hangi günler çalıştığı</b>: 7 bitlik maske,
/// Pazartesi=1 · Salı=2 · Çarşamba=4 · Perşembe=8 · Cuma=16 · Cumartesi=32 · Pazar=64.
/// </summary>
/// <remarks>
/// 🔴 <b>Gün dönüşümünün tek sahibi burasıdır.</b> .NET <see cref="DayOfWeek"/> <b>Pazar=0</b>'dan
/// başlar, bu maske <b>Pazartesi</b>'den — klasik bir sessiz kayma hatası. İki yerde ayrı
/// yazılsaydı "Salı seferi Pazartesi görünür" ve <b>kimse hata almazdı</b> (görünmez sözleşme
/// #23'ün sınıfı). Bu yüzden hem panel, hem DTO, hem de 12.6'daki "sıradaki sefer" hesabı
/// aynı metotlardan geçer.
///
/// 🔴 <b><see cref="Mask"/> = 0 yasaktır</b> (<see cref="IsValid"/>): hiçbir gün çalışmayan bir
/// sefer, panelde <i>duran</i> ama mobilde <i>hiç görünmeyen</i> bir kayıttır — yönetici saati
/// girdiğini sanır, vatandaş hiçbir zaman göremez. Doğrulama komutta yapılır.
///
/// ⚠️ Kodlar (<c>"mon"</c>…<c>"sun"</c>) DTO'ya çıkar, yani <b>kontrattır</b>: değişirse
/// mağazadaki eski sürümler günü tanımaz. Türkçe karşılıkları panelde/mobilde üretilir.
/// </remarks>
public readonly record struct OperatingDays
{
    public const int Monday = 1;
    public const int Tuesday = 2;
    public const int Wednesday = 4;
    public const int Thursday = 8;
    public const int Friday = 16;
    public const int Saturday = 32;
    public const int Sunday = 64;

    /// <summary>Her gün — 12.5 öncesinden kalan bütün seferlerin değeri (davranış değişmedi).</summary>
    public const int Daily = 127;

    /// <summary>Pazartesi–Cuma.</summary>
    public const int Weekdays = Monday | Tuesday | Wednesday | Thursday | Friday;

    /// <summary>Cumartesi–Pazar.</summary>
    public const int Weekend = Saturday | Sunday;

    /// <summary>Maskedeki bit sırası — <b>Pazartesi'den başlar</b>, DTO kodları da bu sıradadır.</summary>
    private static readonly (int Bit, string Code, DayOfWeek Day)[] Days =
    {
        (Monday, "mon", DayOfWeek.Monday),
        (Tuesday, "tue", DayOfWeek.Tuesday),
        (Wednesday, "wed", DayOfWeek.Wednesday),
        (Thursday, "thu", DayOfWeek.Thursday),
        (Friday, "fri", DayOfWeek.Friday),
        (Saturday, "sat", DayOfWeek.Saturday),
        (Sunday, "sun", DayOfWeek.Sunday)
    };

    public int Mask { get; }

    public OperatingDays(int mask) => Mask = mask;

    public static OperatingDays All => new(Daily);

    /// <summary>1–127 aralığı. 0 "hiçbir gün" demektir ve <b>geçersizdir</b>.</summary>
    public bool IsValid => Mask is > 0 and <= Daily;

    public bool RunsDaily => Mask == Daily;
    public bool RunsWeekdaysOnly => Mask == Weekdays;
    public bool RunsWeekendOnly => Mask == Weekend;

    /// <summary>🔴 Pazar=0 kayması tam olarak burada çözülür — başka hiçbir yerde tekrarlanmaz.</summary>
    public static int BitFor(DayOfWeek day) => day switch
    {
        DayOfWeek.Monday => Monday,
        DayOfWeek.Tuesday => Tuesday,
        DayOfWeek.Wednesday => Wednesday,
        DayOfWeek.Thursday => Thursday,
        DayOfWeek.Friday => Friday,
        DayOfWeek.Saturday => Saturday,
        _ => Sunday
    };

    public bool Runs(DayOfWeek day) => (Mask & BitFor(day)) != 0;

    /// <summary>DTO'ya çıkan kodlar, <b>Pazartesi'den Pazar'a</b> sıralı.</summary>
    public IReadOnlyList<string> Codes()
    {
        var codes = new List<string>(7);
        foreach (var (bit, code, _) in Days)
            if ((Mask & bit) != 0)
                codes.Add(code);
        return codes;
    }

    /// <summary>
    /// Kodlardan maske üretir; tanınmayan kod <b>yok sayılır</b> (bilinmeyen değer listeyi
    /// patlatmamalı). Hiçbiri tanınmazsa maske 0 döner — çağıran <see cref="IsValid"/> ile
    /// reddeder.
    /// </summary>
    public static OperatingDays FromCodes(IEnumerable<string>? codes)
    {
        var mask = 0;
        if (codes is null) return new OperatingDays(0);

        foreach (var raw in codes)
        {
            var code = raw?.Trim().ToLowerInvariant();
            foreach (var (bit, known, _) in Days)
                if (string.Equals(code, known, StringComparison.Ordinal))
                    mask |= bit;
        }

        return new OperatingDays(mask);
    }

    /// <summary>Panel formundaki 7 onay kutusunun ürettiği bit listesinden maske.</summary>
    public static OperatingDays FromBits(IEnumerable<int>? bits)
    {
        var mask = 0;
        if (bits is null) return new OperatingDays(0);

        foreach (var bit in bits)
            if (Days.Any(d => d.Bit == bit))
                mask |= bit;

        return new OperatingDays(mask);
    }

    /// <summary>
    /// <paramref name="from"/> gününden başlayarak (o gün <b>dâhil</b>) seferin çalıştığı ilk
    /// güne kaç gün olduğu; hiç çalışmıyorsa <c>null</c>. 12.6'nın "sıradaki sefer"i buna dayanır.
    /// </summary>
    public int? DaysUntilNext(DayOfWeek from)
    {
        if (Mask == 0) return null;

        for (var offset = 0; offset < 7; offset++)
        {
            var day = (DayOfWeek)(((int)from + offset) % 7);
            if (Runs(day)) return offset;
        }

        return null;
    }

    /// <summary>Maskedeki günlerin <see cref="DayOfWeek"/> karşılıkları (Pazartesi'den sıralı).</summary>
    public IReadOnlyList<DayOfWeek> ToDayOfWeeks()
    {
        // ⚠️ `this` bir struct olduğu için lambda içinde kullanılamaz (CS1673) — yerele kopyala.
        var mask = Mask;
        return Days.Where(d => (mask & d.Bit) != 0).Select(d => d.Day).ToList();
    }

    public override string ToString() => Mask.ToString(CultureInfo.InvariantCulture);
}

File: mobile/lib/features/transport/application/operating_days.dart
import 'package:flutter/foundation.dart';

/// Faz 12.6 — bir seferin **hangi günler çalıştığı**; sunucudaki
/// `OperatingDays` değer nesnesinin mobil karşılığı.
///
/// 🔴 **Gün dönüşümünün mobildeki TEK SAHİBİ burasıdır** (görünmez sözleşme
/// #46'nın istemci tarafı). Sunucu maskesi **Pazartesi=1**'den başlar; Dart'ın
/// `DateTime.weekday`'i de **Pazartesi=1 … Pazar=7** olduğu için .NET'teki
/// "Pazar=0" kayması burada *yok* — ama tam bu yüzden tehlikeli: `1 << weekday`
/// ya da `weekday % 7` yazan ikinci bir eşleme **bir gün kaydırır**, ekran
/// açılır, hata vermez ve yalnız gün yanlış olur. Doğru bağıntı tek satırda:
/// `bit = 1 << (weekday - 1)`.
///
/// ⚠️ Kodlar (`"mon"`…`"sun"`) **kontrattır** — sunucu `days` alanında bunları
/// gönderiyor. Türkçe karşılıkları burada üretilir, sunucudan gelmez.
@immutable
class OperatingDays {
  const OperatingDays(this.mask);

  static const int monday = 1;
  static const int tuesday = 2;
  static const int wednesday = 4;
  static const int thursday = 8;
  static const int friday = 16;
  static const int saturday = 32;
  static const int sunday = 64;

  /// Her gün — 12.5 öncesinden kalan bütün seferlerin değeri.
  static const int dailyMask =
      monday | tuesday | wednesday | thursday | friday | saturday | sunday;

  static const int weekdaysMask =
      monday | tuesday | wednesday | thursday | friday;

  static const int weekendMask = saturday | sunday;

  static const OperatingDays daily = OperatingDays(dailyMask);

  /// Kod → bit, **Pazartesi'den Pazar'a** (kontrat sırası).
  static const List<(int, String, String)> _days = [
    (monday, 'mon', 'Pzt'),
    (tuesday, 'tue', 'Sal'),
    (wednesday, 'wed', 'Çar'),
    (thursday, 'thu', 'Per'),
    (friday, 'fri', 'Cum'),
    (saturday, 'sat', 'Cmt'),
    (sunday, 'sun', 'Paz'),
  ];

  final int mask;

  /// Sunucudan gelen kod listesinden maske.
  ///
  /// 🔴 **Boş/eksik liste "hiçbir gün" DEĞİL, "her gün" sayılır.** Alan 12.5'te
  /// additive olarak eklendi; tanımadığımız ya da hiç gelmeyen bir `days`'i
  /// "hiç çalışmıyor" saymak seferi ekrandan **sessizce silerdi** — oysa
  /// sunucunun `runsDaily` varsayılanı `true` ve uç seferleri günlere göre
  /// *elemiyor*, yalnız *bildiriyor* (görünmez sözleşme #46). Şüphede kalınca
  /// **göstermek** doğru yön.
  ///
  /// Tanınmayan kod **yok sayılır** (bir yazım hatası listeyi boşaltmamalı).
  factory OperatingDays.fromCodes(Iterable<String>? codes) {
    if (codes == null) return daily;

    var mask = 0;
    for (final raw in codes) {
      final code = raw.trim().toLowerCase();
      for (final (bit, known, _) in _days) {
        if (code == known) mask |= bit;
      }
    }

    return mask == 0 ? daily : OperatingDays(mask);
  }

  bool get runsDaily => mask == dailyMask;
  bool get runsWeekdaysOnly => mask == weekdaysMask;
  bool get runsWeekendOnly => mask == weekendMask;

  /// 🔴 Pazar kayması **yalnız burada** çözülür — `DateTime.weekday`
  /// Pazartesi=1 … Pazar=7.
  static int bitForWeekday(int weekday) {
    if (weekday < DateTime.monday || weekday > DateTime.sunday) return 0;
    return 1 << (weekday - 1);
  }

  bool runsOnWeekday(int weekday) => (mask & bitForWeekday(weekday)) != 0;

  bool runsOn(DateTime date) => runsOnWeekday(date.weekday);

  /// [fromWeekday] gününden başlayarak (o gün **dâhil**) seferin çalıştığı ilk
  /// güne kaç gün olduğu; hiç çalışmıyorsa `null`.
  ///
  /// ⚠️ Döngü **7 turdur, 6 değil**: yalnız Pazartesi çalışan bir sefere
  /// Pazartesi günü sorulursa cevap 0'dır, Salı sorulursa 6'dır — ama
  /// "bugünü atlayıp bir hafta sonrası" hesabı [DepartureTimes] tarafında
  /// 7. ofsetle yapılır (bkz. `nextAmong`).
  int? daysUntilNext(int fromWeekday) {
    if (mask == 0) return null;

    for (var offset = 0; offset < 7; offset++) {
      // Dart'ta gün 1..7 → 0 tabanına in, mod al, geri çık.
      final weekday = ((fromWeekday - 1 + offset) % 7) + 1;
      if (runsOnWeekday(weekday)) return offset;
    }

    return null;
  }

  /// DTO'ya çıkan kodlar, Pazartesi'den Pazar'a sıralı.
  List<String> get codes => [
    for (final (bit, code, _) in _days)
      if ((mask & bit) != 0) code,
  ];

  /// "Pzt · Çar · Cum" — kısa gün adları.
  List<String> get shortNames => [
    for (final (bit, _, name) in _days)
      if ((mask & bit) != 0) name,
  ];

  /// Kartta gösterilen rozet metni.
  ///
  /// "Her gün" bilinçli olarak **yazılır**: rozetsiz bir saat "gün bilgisi yok"
  /// mu yoksa "her gün" mü belli olmaz. Çağıran ekran her gün çalışan seferde
  /// rozeti gizlemeyi seçebilir ([runsDaily] ile) — karar orada verilir, burada
  /// metin her zaman doğrudur.
  String get label {
    if (runsDaily) return 'Her gün';
    if (runsWeekdaysOnly) return 'Hafta içi';
    if (runsWeekendOnly) return 'Hafta sonu';
    final names = shortNames;
    if (names.isEmpty) return 'Sefer günü belirtilmemiş';
    return names.join(' · ');
  }

  /// Ekran okuyucu için açık metin — "Pzt · Çar" kısaltması sesli okunduğunda
  /// anlaşılmaz.
  String get semanticsLabel {
    if (runsDaily) return 'her gün';
    if (runsWeekdaysOnly) return 'hafta içi her gün';
    if (runsWeekendOnly) return 'hafta sonu';
    const long = {
      'Pzt': 'Pazartesi',
      'Sal': 'Salı',
      'Çar': 'Çarşamba',
      'Per': 'Perşembe',
      'Cum': 'Cuma',
      'Cmt': 'Cumartesi',
      'Paz': 'Pazar',
    };
    final names = shortNames;
    if (names.isEmpty) return 'sefer günü belirtilmemiş';
    return names.map((name) => long[name] ?? name).join(', ');
  }

  /// "Pzt" — [weekday] Dart günü (1..7). Sıradaki sefer satırında kullanılır.
  static String shortNameOfWeekday(int weekday) {
    for (final (bit, _, name) in _days) {
      if (bit == bitForWeekday(weekday)) return name;
    }
    return '';
  }

  @override
  bool operator ==(Object other) => other is OperatingDays && other.mask == mask;

  @override
  int get hashCode => mask.hashCode;

  @override
  String toString() => 'OperatingDays($label)';
}
